wren-modeling-rs/core/src/mdl/utils.rs

```diff
@@ -89,105 +86,97 @@ pub fn create_wren_calculated_field_expr(
-    expr: &String,
+    expr: &str,
     model: Arc<Model>,
     analyzed_wren_mdl: Arc<AnalyzedWrenMDL>,
+    session_state: SessionStateRef,
 ) -> Result<Expr> {
-    let context_provider = RemoteContextProvider::new(&analyzed_wren_mdl.wren_mdl)?;
-    create_expr_for_model(
-        expr,
-        model,
-        DynamicContextProvider::new(Box::new(context_provider)),
+    let dataset = Dataset::Model(model);
+    let schema = dataset.to_remote_schema(
+        Some(analyzed_wren_mdl.wren_mdl().get_register_tables()),
+        Arc::clone(&session_state),
+    )?;
+    let session_state = session_state.read();
+    session_state.create_logical_expr(
+        qualified_expr(expr, &schema, &session_state)?.as_str(),
+        &schema,
     )
 }
 
 /// Create the Logical Expr for the remote column.
-/// Use [RemoteContextProvider] to provide the context for the remote column.
 pub(crate) fn create_wren_expr_for_model(
-    expr: &String,
+    expr: &str,
     model: Arc<Model>,
-    analyzed_wren_mdl: Arc<AnalyzedWrenMDL>,
+    session_state: SessionStateRef,
 ) -> Result<Expr> {
-    let context_provider = WrenContextProvider::new(&analyzed_wren_mdl.wren_mdl)?;
-    let wrapped = format!(
-        "select {} from {}.{}.{}",
-        expr,
-        analyzed_wren_mdl.wren_mdl().catalog(),
-        analyzed_wren_mdl.wren_mdl().schema(),
-        &model.name
-    );
-    let parsed = Parser::parse_sql(&GenericDialect {}, &wrapped).unwrap();
-    let statement = &parsed[0];
-    debug!("Statement: {:?}", statement.to_string());
-
-    let sql_to_rel = SqlToRel::new(&context_provider);
-    let plan = sql_to_rel.sql_statement_to_plan(statement.clone())?;
-
-    match plan {
-        LogicalPlan::Projection(projection) => Ok(projection.expr[0].clone()),
-        _ => internal_err!("Unexpected plan type: {:?}", plan),
-    }
+    let dataset = Dataset::Model(model);
+    let schema = dataset.to_qualified_schema()?;
+    let session_state = session_state.read();
+    session_state.create_logical_expr(
+        qualified_expr(expr, &schema, &session_state)?.as_str(),
+        &schema,
+    )
 }
 
-/// Create the Logical Expr for the column belong to the model according to the context provider
-pub(crate) fn create_expr_for_model(
-    expr: &String,
-    model: Arc<Model>,
-    context_provider: DynamicContextProvider,
-) -> Result<Expr> {
-    let wrapped = format!("select {} from {}", expr, &model.table_reference);
-    let parsed = Parser::parse_sql(&GenericDialect {}, &wrapped)?;
-    let statement = &parsed[0];
-
-    let sql_to_rel = SqlToRel::new(&context_provider);
-    let plan = sql_to_rel.sql_statement_to_plan(statement.clone())?;
-    match plan {
-        LogicalPlan::Projection(projection) => Ok(projection.expr[0].clone()),
-        _ => internal_err!("Unexpected plan type: {:?}", plan),
-    }
+/// Add the table prefix for the column expression if it can be resolved by the schema.
+fn qualified_expr(
```

Comment:
I have filed an issue https://github.com/apache/datafusion/issues/11551 for it. If the behavior is improved, we can remove this function.